utils/src/main/kotlin/com/exactpro/th2/lwdataprovider/MessageSearcher.kt

```diff
@@ -0,0 +1,156 @@
+            sessionGroup = sessionGroup?.ifBlank { sessionAlias } ?: sessionAlias,
+            sessionAlias = sessionAlias,
+            direction = direction
+        )
+        K_LOGGER.info { "Search the last or null, base request: ${requestBuilder.toJson()}, start time: $now, end time: $end" }
+
+        return withCancellation {
+            generateSequence {
+                if (to.isBefore(end)) {
+                    return@generateSequence null
+                }
+
+                from = to
+                to = from.minus(requestInterval)
+
+                service.searchMessageGroups(
+                    requestBuilder.apply {
+                        startTimestamp = from.toTimestamp()
+                        endTimestamp = to.toTimestamp()
+                    }.build().also {
+                        K_LOGGER.debug { "Request from: $from, to: $to" }
+                    }
+                )
+            }.flatMap(Iterator<MessageSearchResponse>::asSequence)
+                .onEach { K_LOGGER.trace { "Message: ${it.toJson()}" } }
+                .firstOrNull(filter::apply)
+        }
+    }
+
+    private fun createRequestBuilder(
+        book: String,
+        sessionGroup: String,
+        sessionAlias: String,
+        direction: Direction,
+    ) = MessageGroupsSearchRequest.newBuilder().apply {
+        bookIdBuilder.name = book
+        addMessageGroup(MessageGroupsSearchRequest.Group.newBuilder().setName(sessionGroup))
+        addStream(
+            MessageStream.newBuilder()
+                .setName(sessionAlias)
+                .setDirection(direction)
+        )
+        addResponseFormats(BASE64_FORMAT)
+        searchDirection = TimeRelation.PREVIOUS
+    }
+
+    companion object {
+        private val K_LOGGER = KotlinLogging.logger { }
+        private const val BASE64_FORMAT = "BASE_64"
+
+        @JvmField
+        val DEFAULT_REQUEST_INTERVAL: Duration = Duration.ofDays(1)
+
+        @JvmOverloads
+        fun MessageSearcher.with(
+            requestInterval: Duration = DEFAULT_REQUEST_INTERVAL,
+            defaultBook: String = "",
+            defaultSessionGroup: String? = null,
+            defaultSessionAlias: String = "",
+        ) = MessageSearcher(service, requestInterval, defaultBook, defaultSessionGroup, defaultSessionAlias)
+
+        fun MessageSearcher.withRequestInterval(requestInterval: Duration) =
+            with(requestInterval, defaultBook, defaultSessionGroup, defaultSessionAlias)
+
+        fun MessageSearcher.withBook(defaultBook: String) =
+            with(requestInterval, defaultBook, defaultSessionGroup, defaultSessionAlias)
+
+        fun MessageSearcher.withSessionGroup(defaultSessionGroup: String) =
+            with(requestInterval, defaultBook, defaultSessionGroup, defaultSessionAlias)
+
+        fun MessageSearcher.withSessionAlias(defaultSessionAlias: String) =
+            with(requestInterval, defaultBook, defaultSessionGroup, defaultSessionAlias)
```

Comment:
Just for your information: this might look good (even great) in Kotlin but in Java this will be static method in `MessageSearcher.Companion` class. Don't think this will be useful in Java in this format. I think we should either create a builder class for searcher or move this method to the `MessageSearcher` class